src/executor/stack/executor.rs

```diff
@@ -838,11 +880,17 @@ impl<'config, 'precompiles, S: StackState<'config>, P: PrecompileSet>
 			}
 		}
 
-		let code = self.code(code_address);
-
 		self.enter_substate(gas_limit, is_static);
 		self.state.touch(context.address);
 
+		#[cfg(feature = "with-substrate")]
+		if let Err(e) =
+			self.record_external_operation(crate::ExternalOperation::AddressCodeRead(code_address))
+		{
+			let _ = self.exit_substate(StackExitKind::Failed);
+			return Capture::Exit((ExitReason::Error(e), Vec::new()));
+		}
+		let code = self.code(code_address);
```

Comment:
Do we need to record the `AccountBasicRead` operation gas here for the following transfer? You have already done this for the increase nonce, transfer touch the account info storage as well.

Reply:
And this, See https://github.com/rust-blockchain/evm/pull/170/files#r1235004396

Reply:
> transfer touch the account info storage as well

where? in the case of frontier's `StackState` impl, `touch` does not actually read the storage

Reply:
- https://github.com/rust-blockchain/evm/blob/master/src/executor/stack/executor.rs#LL852C3-L852C3

Sorry for the unclear comment, I mean the balance transfer operation in the `transact_create` and `transact_call`. The account balance is stored together with the nonce in the substrate storage.

Reply:
> The account balance is stored together with the nonce in the substrate storage. I am not sure if we need to add another AccountBasicRead cost, it seems record once when increasing the account nonce is enough.

Oh I see thank you for explaining. We aim to be as agnostic as possible of the `StackState` implementation - in Substrate what you mentioned might be true by default (but not necessarily). Same applies for non-substrate chains using this crate.

Anyway you are right, I will add an additional `AccountBasicRead` before transfer